cpp/src/arrow/engine/exec_plan.h

```diff
@@ -0,0 +1,131 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/compute/type_fwd.h"
+#include "arrow/type_fwd.h"
+#include "arrow/util/future.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/mutex.h"
+#include "arrow/util/optional.h"
+#include "arrow/util/type_fwd.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+namespace engine {
+
+class ExecNode;
+class QueryNode;
+class QueryPlan;
+
+class ARROW_EXPORT ExecPlan {
+ public:
+  using NodeVector = std::vector<ExecNode*>;
+
+  virtual ~ExecPlan() = default;
+
+  /// The query plan this ExecPlan is an instance of
+  const QueryPlan& query_plan() const { return *query_plan_; }
+
+  compute::ExecContext* context() { return context_; }
+
+ protected:
+  friend class QueryPlan;
+
+  ExecPlan() = default;
+
+  const std::shared_ptr<const QueryPlan> query_plan_;
+  compute::ExecContext* context_;
+};
+
+class ARROW_EXPORT ExecNode {
+ public:
+  using NodeVector = std::vector<ExecNode*>;
+
+  virtual ~ExecNode();
+
+  /// The query node this ExecNode is an instance of
+  const QueryNode& query_node() const { return *query_node_; }
+
+  /// This node's exec plan
+  ExecPlan* plan() { return plan_; }
+
+  /// Transfer input batch to ExecNode
+  ///
+  /// When all inputs are received for a given batch_index, the batch is ready
+  /// for execution.
+  Status InputReceived(int32_t input_index, int32_t batch_index,
+                       compute::ExecBatch batch);
```

Comment:
What is "batch_index"?

Reply:
The index of the batch in the batch stream.

Reply:
I don't think this parameter is important, at least at this stage. Could you clarify what "When all inputs are received for a given batch_index, the batch is ready for execution." means? In general, inputs from different parents won't necessarily correspond to each other.

For example, suppose that we are joining a 1,000,000 row input with a 100,000 row input. One input might yield 100 batches while the other might only yield 5 or 10. They may be of differing lengths, in an unpredictable order (since scans won't necessarily yield a deterministic order), and the batches won't correspond to each other.

Reply:
>  In general, inputs from different parents won't necessarily correspond to each other.

Ah, I see. This is a misunderstanding on my part, then. Sorry.